Database Management

brandonshearin opened the pull request:
## Description

<!--- Describe your changes in detail -->
This changeset adds a new page underneath "administration" that enables a user to obliterate their data. 

**3/1 UPDATE** 
the api contract has changed slightly to accommodate more flexibility in deleting asset group selectors.  the request body used to have a boolean field named `deleteHighValueSelectors`, which has been changed to the name `deleteAssetGroupSelectors` and it accepts a list of ints (corresponding to asset group id's)

the ui has been updated with 5 checkboxes instead of 4:
<img width="445" alt="Screenshot 2024-03-01 at 10 49 25 AM" src="https://github.com/SpecterOps/BloodHound/assets/23143242/d375f4b3-1f92-4f2f-a345-6925c87b57bb">



api changes: 
- added a new "WipeDB" permission to the "administrator" role 
- added a new endpoint `POST api/v2/clear-database` 
- added 3 new methods to the bloodhound db interface to delete data: **DeleteAllAssetGroupSelectorsForAssetGroups()** , **DeleteAllDataQuality()**, and **DeleteAllFileUploads**
- add swagger docs for new endpoint (this is going to get deleted after cody's PR so i'll keep track of that)

ui changes:
- new `<DatabaseManagement />` and `<ConfirmationDialog/>` components

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
give a user the ability to light fire to their data

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

unit tests have been added to cover UI changes, and unit and integration tests have been added for the api handler

## Screenshots (if appropriate):


## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

mistahj67 commented on cmd/api/src/auth/permission.go:
```diff
@@ -95,5 +98,7 @@ func Permissions() PermissionSet {
 		SavedQueriesRead:  model.NewPermission("saved_queries", "Read"),
 		SavedQueriesWrite: model.NewPermission("saved_queries", "Write"),
+
+		WipeDB: model.NewPermission("wipedb", "wipedb"),
```
Style nit here but it seems like we follow a patter of lowercase for authority and then uppercase for name and i'm not sure if we wanted to make the authority generalized such as:
`db` and then for name `Wipe` perhaps?

elikmiller replied:
I agree that the naming here could be improved

urangel commented on cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:
```diff
@@ -0,0 +1,307 @@
+    };
+
+    return (
+        <ContentPage title='Clear BloodHound data'>
```
We should capitalize _data_ here to align with the casing in the other page titles.

urangel commented on cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:
```diff
@@ -0,0 +1,307 @@
+                            </Box>
+                        ) : null}
+
+                        {state.mutationError ? (
+                            <Box color={theme.palette.error.main} display='flex' alignItems='center' gap='0.3rem'>
+                                <FontAwesomeIcon icon={faCircleXmark} />
+                                <FormHelperText sx={{ marginTop: '1.5px', color: theme.palette.error.main }}>
+                                    There was an error processing your request.
+                                </FormHelperText>
+                            </Box>
+                        ) : null}
```
We use the MUI Alert components in other places in the app. This looks like a good place to use an <Alert/> so that these kinds of messages look consistent.

urangel commented on cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:
```diff
@@ -0,0 +1,307 @@
+
+    return (
+        <ContentPage title='Clear BloodHound data'>
+            <div>
```
It would be nice to swap this out for an MUI \<Box/> so that it applies the generated MUI css classes. It is effectively the same right now but if we ever start introducing any theming (like dark mode 👀 ) it will be easier to have MUI handle sweeping style changes rather than having to update this div manually. Wherever possible, try to use MUI components.

urangel commented on cmd/api/src/database/agi.go:
```diff
@@ -41,9 +41,11 @@ func (s *BloodhoundDB) CreateAssetGroup(ctx context.Context, name, tag string, s
 	)
 
-	return assetGroup, s.AuditableTransaction(ctx, auditEntry, func(tx *gorm.DB) error {
+	err := s.AuditableTransaction(ctx, auditEntry, func(tx *gorm.DB) error {
 		return CheckError(tx.Create(&assetGroup))
 	})
+
+	return assetGroup, err
```
This is a nice clean up for readability 👍 What do you think about also moving the err declaration into the var block above? The same can be done for the change on line 180 if you decide to do so.

urangel reviewed (commented):
Good job making both the API and UI changes on this! Also, thanks for all the tests you included. This feature will be a big quality of life improvement. Most, if not all, of the comments I left were minor consistency nits but the logic checks out. 

On a final note, I understand that we want to make it difficult to accidentally delete your data and that is why we ask for typing the phrase in but it felt a bit too cumbersome to me. If nothing else, maybe we can make the phrase matching case insensitive? I enjoyed the way legacy BloodHound would swap the confirm delete button with the cancel button in case one was spamming clicks. Just my two cents 🤷

elikmiller commented:
I don't see our swagger docs updated with the new endpoint, those updates need to be included.

elikmiller commented:
The ticket calls out that we should "initiate a new analysis" after deleting data - I think this piece is missing.

mistahj67 reviewed (changes requested):
We discussed in DM but the permissions addition requires a migration to be added to create the permission as well as to add the permission to the admin role.

mistahj67 commented on cmd/api/src/api/v2/database_management_integration_test.go:
```diff
@@ -0,0 +1,75 @@
+
+					return nil
+				})
+
```
we should assert / check the read transaction has no error right? also might be good to verify that the data is first in the db and then the data is no longer in the db by doing a get around the delete?

benwaples commented on cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:
```diff
@@ -0,0 +1,307 @@
+        case 'open_dialog': {
+            const { collectedGraphData, dataQualityHistory, fileIngestHistory, highValueSelectors } = state;
+            if (
+                [collectedGraphData, dataQualityHistory, fileIngestHistory, highValueSelectors].filter(Boolean)
```
Loving the separation of concerns here in this file!

nit: Could we pull this out to a named variable to help give context as to what we are checking? Its not very obvious at first and takes a couple reads to see whats going on!

Maybe something like 
```js
const noSelection = collectedGraphData, dataQualityHistory, fileIngestHistory, highValueSelectors].filter(Boolean).length === 0

if (noSelection) {
  ...
}

```

benwaples commented on cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx:
```diff
@@ -0,0 +1,97 @@
+                </Box>
+            </DialogContent>
+            <DialogActions>
+                <Button onClick={handleClose}>Cancel</Button>
```
minor ux thing: IMO it would be nice to clear the error and input state on cancel as well!

elikmiller replied:
The Cancel and Confirm buttons should have two different colors

benwaples commented on cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx:
```diff
@@ -0,0 +1,97 @@
+                    </Typography>
+                    <Typography variant='body1'>Please input the phrase prior to click confirm.</Typography>
+                    <TextField
+                        placeholder='Please delete my data'
```
We could reuse the `confirmationText ` variable here so if we need to change it in the future then we only need to change it in one place!!

elikmiller commented on cmd/api/src/api/registration/v2.go:
```diff
@@ -131,6 +131,8 @@ func NewV2API(cfg config.Configuration, resources v2.Resources, routerInst *rout
 		routerInst.GET("/api/v2/features", resources.GetFlags).RequirePermissions(permissions.AppReadApplicationConfiguration),
 		routerInst.PUT("/api/v2/features/{feature_id}/toggle", resources.ToggleFlag).RequirePermissions(permissions.AppWriteApplicationConfiguration),
 
+		routerInst.POST("/api/v2/database-management", resources.HandleDatabaseManagement).RequirePermissions(permissions.WipeDB),
```
My opinion is that this endpoint should be renamed. To me "database management" implies a lot more beyond clearing data. Maybe "/api/v2/clear-database"?

elikmiller commented on cmd/api/src/api/v2/database_management_integration_test.go:
```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Specter Ops, Inc.
```
Should be 2024 for new files

elikmiller commented on cmd/api/src/api/v2/database_management.go:
```diff
@@ -0,0 +1,155 @@
+	}
+
+	// delete custom high value selectors
+	if payload.HighValueSelectors {
```
What is the expectation if both `payload.CollectedGraphData` and `payload.HighValueSelectors` are true but an error occurs inside this block? Do we end up with only some data being deleted instead of all requested data deleted?

elikmiller replied:
Also, it looks like the custom selector is deleted but the nodes themselves remained tagged as part of the asset group. Is that expected behavior @StephenHinck ?

brandonshearin replied:
to your first question, if a user requests to delet emultiple types of data, and gets a partial failure, maybe we should surface that to the UI?

so if i delete graph data and file ingest history, and file ingest history fails, the user would see on the UI something like "we encountered an error deleting file ingest history. please try again."

elikmiller commented on cmd/api/src/api/v2/database_management.go:
```diff
@@ -0,0 +1,155 @@
+		}
+	}
+
+	if err := s.DB.AppendAuditLog(request.Context(), model.AuditEntry{
```
Flagging @juggernot325 since he is up-to-date with all things audit logging. Can you take a look here and validate that this is the appropriate way to create audit logs in this scenario?

juggernot325 replied:
I know we discussed this while you were working on it and decided on generating the audit logs in this method. After looking at the PR as a whole I think we may want to go a different route. One thing that I think is big concern is that if there are any errors in the above operations there will not be an audit log generated at all, so that should be addressed.

I actually think this is going to need a novel implementation of audit logging given the complexity of this operation. Let's meet up and discuss some options.

juggernot325 replied:
The latest iteration of this looks good to me! It should successfully capture the intent audit log and capture actual results.

elikmiller commented on cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx:
```diff
@@ -0,0 +1,97 @@
+                        helperText={
+                            error ? (
+                                <>
+                                    <FontAwesomeIcon icon={faCircleXmark} /> Please input the phrase prior to clicking
```
We should clear this message once the user types the appropriate phrase. It should also re-appear if the phrase is subsequently modified and is no longer matching.

brandonshearin commented on cmd/api/src/database/db.go:
```diff
@@ -138,22 +138,46 @@ type Database interface {
 	DeleteSavedQuery(id int) error
 	SavedQueryBelongsToUser(userID uuid.UUID, savedQueryID int) (bool, error)
+	DeleteAssetGroupSelectors(ctx context.Context, assetGroupId int) error
+	Transaction(fn func(db *BloodhoundDB) error) error
+	Begin() *BloodhoundDB
+	Commit() error
```
i removed this code. it was leftover from something i was trying that was testing related, but i ended up not going that direction

brandonshearin commented:
> We discussed in DM but the permissions addition requires a migration to be added to create the permission as well as to add the permission to the admin role.

appreciate the guidance on this one.  i got this working now-- i added a new migration file titled v5.7.0.sql and added the sql to create the db.Wipe permission and then associated that permission with the Admin role

brandonshearin commented:
> The ticket calls out that we should "initiate a new analysis" after deleting data - I think this piece is missing.

added this to the code block for deleting graph data and for deleting custom high value selectors

mistahj67 commented:
> I don't see our swagger docs updated with the new endpoint, those updates need to be included.

Might be worth coordinating with @codydbentley since he's knee deep in refactoring swagger currently

brandonshearin commented on cmd/api/src/api/v2/database_wipe.go:
```diff
@@ -0,0 +1,161 @@
+		}
+
+		// if succesful, kick off analysis
+		s.TaskNotifier.RequestAnalysis()
```
got it!

brandonshearin commented on cmd/api/src/api/v2/database_wipe_integration_test.go:
```diff
@@ -0,0 +1,89 @@
+	lab.Pack(harness, computer3)
+
+	lab.NewSpec(t, harness).Run(
+		lab.TestCase("the endpoint deletes graph data", func(assert *require.Assertions, harness *lab.Harness) {
```
i added a lot of unit test coverage to this function to cover more bases.  our integration test suite like lab doesn't support writing a test like "create some high value selectors in the db, call the wipe database endpoint, and then assert that there are no high value selectors in the db."

i tried to follow this approach, but what i was running into was a race condition in the test.  i would setup some fixtures in lab to create some data, i.e. some high value selectors, but asset group selectors depend on a FK relationship to an asset group, so I made a fixture to create the asset group first, and then a selector.

some really wonky behavior happened where the asset group record would be created in memory in gorm but wouldn't be committed to the db, so there was a race condition when trying to create the selector.  creating the selector usually threw a FK constraint violation because the asset group wasn't actually committed to the db yet.

it was too burdensome to try to get something to work for the integration test, so i just unit tested instead!

brandonshearin commented on cmd/api/src/database/dataquality.go:
```diff
@@ -182,3 +183,30 @@ func (s *BloodhoundDB) GetAzureDataQualityAggregations(start time.Time, end time
+func (s *BloodhoundDB) DeleteAllDataQuality() error {
+
+	errADQualityAggregations := CheckError(
+		s.db.Exec("DELETE FROM ad_data_quality_aggregations"),
```
sure can! done

mistahj67 commented on cmd/api/src/database/dataquality.go:
```diff
@@ -182,3 +183,11 @@ func (s *BloodhoundDB) GetAzureDataQualityAggregations(start time.Time, end time
 }
+
+func (s *BloodhoundDB) DeleteAllDataQuality(ctx context.Context) error {
+	err := CheckError(
+		s.db.WithContext(ctx).Exec("DELETE FROM ad_data_quality_aggregations; DELETE FROM ad_data_quality_stats; DELETE FROM azure_data_quality_aggregations; DELETE FROM azure_data_quality_stats;"),
+	)
+
+	return err
```
```suggestion
	return CheckError(
		s.db.WithContext(ctx).Exec("DELETE FROM ad_data_quality_aggregations; DELETE FROM ad_data_quality_stats; DELETE FROM azure_data_quality_aggregations; DELETE FROM azure_data_quality_stats;"),
	)
```

mistahj67 commented on cmd/api/src/go.mod:
```diff
@@ -47,6 +47,7 @@ require (
 	github.com/unrolled/secure v1.13.0
 	github.com/zenazn/goji v1.0.1
 	go.uber.org/mock v0.2.0
+	go.uber.org/zap v1.13.0
```
What are these go.uber.org/zap for?

mistahj67 commented on cmd/ui/src/views/Administration/Administration.tsx:
```diff
@@ -48,6 +50,11 @@ const Administration: React.FC = () => {
                     component: QA,
                 },
+                {
+                    label: 'Database Management',
```
Did we want to keep this as Database Management ? 🤔 Probably should check with product

brandonshearin replied:
yep! this is the name specified in the acceptance criteria

mistahj67 reviewed (commented):
Backend logic LGTM! Thanks for your patience with all of my feedback!

mistahj67 reviewed (approved):
Approving to remove my requested changes blocker but definitely get a more seasoned front end approval before merging!

maffkipp commented on cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:
```diff
@@ -0,0 +1,293 @@
+// Copyright 2024 Specter Ops, Inc.
```
I think it might be helpful to break the reducer, hook, and component out into separate files here

mistahj67 commented on cmd/api/src/database/agi.go:
```diff
@@ -204,6 +213,13 @@ func (s *BloodhoundDB) DeleteAssetGroupSelector(ctx context.Context, selector mo
 
+func (s *BloodhoundDB) DeleteAssetGroupSelectorsForAssetGroup(ctx context.Context, assetGroupId int) error {
+	return CheckError(
+		s.db.Where("asset_group_id = ?", assetGroupId).
```
We need the `WithContext(ctx)` here on this db call

brandonshearin replied:
gotcha

elikmiller commented on cmd/api/src/api/v2/database_wipe.go:
```diff
@@ -0,0 +1,243 @@
+	DeleteHighValueSelectors bool `json:"deleteHighValueSelectors"`
+	DeleteFileIngestHistory  bool `json:"deleteFileIngestHistory"`
+	DeleteDataQualityHistory bool `json:"deleteDataQualityHistory"`
+	AssetGroupId             int  `json:"assetGroupId"`
```
If the purpose of this endpoint is to "delete high value selectors" - why is the onus on the requestor to provide the correct asset group ID? Can the service not derive the asset group ID for the appropriate asset group?

StephenHinck replied:
I forgot to consider in the AC for this ticket that multiple groups exist (and will become more prevalent in the future).

Would it make more sense to leave it this way, change `DeleteHighValueSelectors` to `DeleteGroupSelectors` and then move `AssetGroupId` to an array/list for improved long-term supportability (with an 'all' option)? I recognize this would be a change to AC and scope, and I am happy to discuss it.

elikmiller replied:
@StephenHinck it is your feature/call to make. I think that what @brandonshearin has implemented satisfies your original AC but if that is no longer desired then we will need time to discuss and make adjustments. Consider that once this changeset is in and released it will be harder to make changes to the API contract.

elikmiller commented on cmd/api/src/api/v2/database_wipe.go:
```diff
@@ -0,0 +1,243 @@
+		return
+	}
+
+	// return `BadRequest` if request specifies `high value selectors` without an `asset group ID`
+	if payload.DeleteHighValueSelectors && payload.AssetGroupId == 0 {
+		api.WriteErrorResponse(
+			request.Context(),
+			api.BuildErrorResponse(http.StatusBadRequest, "please provide an assetGroupId to delete", request),
+			response,
+		)
+		return
+	}
```
Assuming the answer to my previous question is "the service should not need the requestor to provide an asset group ID" then we can remove this validation

brandonshearin replied:
removed!
